### DISCUSSION TITLE
Is there a list of native functions somewhere?

### DISCUSSION THREAD

**Discussion by HJHughJanus** [wopss/RED4ext#10] [Q&A]
Hello, I am currently trying to figure out if its the right time for getting into cyberpunk modding, or if I should wait another year or so.
What would help me is a list of native functions currently available, so I can know what would be possible already.
Does anything like that exist?

BR
HJ

**Comment by wopss** [comment]
Hello,

Yes something like this exists, I usually dump it for every patch, we currently have it on the modding Discord (https://discord.gg/Epkq79kd96), this is the link for the dump https://cdn.discordapp.com/attachments/794613856948322344/827304684582600754/dump_rtti.zip (for version 1.2). It changes with each patch but you can also check this message for new updates https://discord.com/channels/717692382849663036/807165519018459156/808700512173162576.

**Reply by HJHughJanus** [reply]
I dont have a discord account and dont plan to get one, so its neat that there is a download link. Thanks.
So did I understand correctly that you are still finding new native functions, thus the list grows larger with every "release"?

**Reply by wopss** [reply]
Yes, it is correct, we can find new natives if they add them in their scripts. Every update can have new natives or remove old ones, so we just run a dump of the game's scripting RTTI every update to have an up-to-date list of natives.

This is the last dump for 1.21 (https://cdn.discordapp.com/attachments/794613856948322344/831856683709104128/rtti_dump.zip) which was released few hours ago.

**Reply by HJHughJanus** [reply]
Why would you remove natives?
Only because they are not used in scripts anymore, does not mean they have been deleted, or did I miss the point?

**Reply by wopss** [reply]
If they are removed, they are not present in the scripts anymore. So trying to call them will  do nothing.

**Reply by PhantomGamers** [reply]
> Why would you remove natives?
> Only because they are not used in scripts anymore, does not mean they have been deleted, or did I miss the point?

"update" is referring to game updates, not updates to RED4Ext. Native functions can be added and removed in the binary of the game itself.

**Reply by wopss** [reply]
> > Why would you remove natives?
> > Only because they are not used in scripts anymore, does not mean they have been deleted, or did I miss the point?
> 
> "update" is referring to game updates, not updates to RED4Ext. Native functions can be added and removed in the binary of the game itself.

This, thank you for clarification!

**Reply by HJHughJanus** [reply]
I see, thanks for making it clear.
So I suppose it is better to wait for next few patches before starting modding right now (because used functions could be removed).

**Reply by wopss** [reply]
> I see, thanks for making it clear.
> So I suppose it is better to wait for next few patches before starting modding right now (because used functions could be removed).

I don't think that you should wait if you want to start modding, we don't know when they will stop dropping patches for Cyberpunk. Functions do get deleted, but most of the times they are not that important and if they are removing a functions you can rewrite it C++ or Lua, so that would not be a problem.

**Reply by HJHughJanus** [reply]
OK. Is there an example mod somewhere, so I can look on how to do basic things?
For example, I would want to iterate through all the loaded NPCs and increase their health.
Can such things be done? Because all I can find in the dump is PrintHealth.

**Reply by wopss** [reply]
> OK. Is there an example mod somewhere, so I can look on how to do basic things?
> For example, I would want to iterate through all the loaded NPCs and increase their health.
> Can such things be done? Because all I can find in the dump is PrintHealth.

I think it should be possible, I saw someone that could toggle things on and off, like lights, vending machines, etc.. You can start by looking at some examples in [RED4ext.SDK](https://github.com/WopsS/RED4ext.SDK/tree/master/examples) project and look in the dump to see how to get the NPCs.

**Reply by HJHughJanus** [reply]
I couldnt find anything useful in the examples. I also checked the last dump again and didnt find any function for setting NPC health. I think I will wait a little more before diving into modding Cyberpunk 2077.

**Reply by HJHughJanus** [reply]
I have now set up the project in VS2019 and loaded your examples. But in the dependencies there are a thousand files - which one do I search for the functions?
In Script Hook (Alexander Blade's) there is always one file called natives.h and this one holds all the functions, so you know if it isnt there, it does not exist. Which file do I have to search in Red4Ext?

**Reply by wopss** [reply]
> 
> 
> I have now set up the project in VS2019 and loaded your examples. But in the dependencies there are a thousand files - which one do I search for the functions?
> In Script Hook (Alexander Blade's) there is always one file called natives.h and this one holds all the functions, so you know if it isnt there, it does not exist. Which file do I have to search in Red4Ext?

Unfortunately I did not create a file like that. For now I would say to get the dumped types with all the functions and variables from here https://cdn.discordapp.com/attachments/794613856948322344/836933371295694889/rtti_dump_1.22_text.zip and look in them for your functions.

**Reply by HJHughJanus** [reply]
How does the SDK know if what I am using is valid?

**Reply by wopss** [reply]
It tries to retrieve the function from the global RTTI, if the function does not exist the return value is nullptr.
